cpp/include/patterns/transform_reduce_e.cuh

```diff
@@ -0,0 +1,220 @@
+                                          EdgeOp>()
+                           .compute(row,
+                                    col,
+                                    weight,
+                                    *(adj_matrix_row_value_input_first + row_offset),
+                                    *(adj_matrix_col_value_input_first + col_offset),
+                                    e_op);
+      e_op_result_sum = plus_edge_op_result(e_op_result_sum, e_op_result);
+    }
+    idx += gridDim.x * blockDim.x;
+  }
+
+  e_op_result_sum =
+    block_reduce_edge_op_result<e_op_result_t,
+                                transform_reduce_e_for_all_low_out_degree_block_size>()
+      .compute(e_op_result_sum);
+  if (threadIdx.x == 0) { *(block_result_first + blockIdx.x) = e_op_result_sum; }
+}
+
+}  // namespace detail
+
+/**
+ * @brief Iterate over the entire set of edges and reduce @p edge_op outputs.
+ *
+ * This function is inspired by thrust::transform_reduce().
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam AdjMatrixRowValueInputIterator Type of the iterator for graph adjacency matrix row
+ * input properties.
+ * @tparam AdjMatrixColValueInputIterator Type of the iterator for graph adjacency matrix column
+ * input properties.
+ * @tparam EdgeOp Type of the binary (or ternary) edge operator.
+ * @tparam T Type of the initial value.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param adj_matrix_row_value_input_first Iterator pointing to the adjacency matrix row input
+ * properties for the first (inclusive) row (assigned to this process in multi-GPU).
+ * `adj_matrix_row_value_input_last` (exclusive) is deduced as @p adj_matrix_row_value_input_first +
+ * @p graph_view.get_number_of_adj_matrix_local_rows().
+ * @param adj_matrix_col_value_input_first Iterator pointing to the adjacency matrix column input
+ * properties for the first (inclusive) column (assigned to this process in multi-GPU).
+ * `adj_matrix_col_value_output_last` (exclusive) is deduced as @p adj_matrix_col_value_output_first
+ * + @p graph_view.get_number_of_adj_matrix_local_cols().
+ * @param e_op Binary (or ternary) operator takes *(@p adj_matrix_row_value_input_first + i), *(@p
+ * adj_matrix_col_value_input_first + j), (and optionally edge weight) (where i and j are row and
+ * column indices, respectively) and returns a transformed value to be reduced.
+ * @param init Initial value to be added to the transform-reduced input vertex properties.
+ * @return T Reduction of the @p edge_op outputs.
+ */
+template <typename GraphViewType,
+          typename AdjMatrixRowValueInputIterator,
+          typename AdjMatrixColValueInputIterator,
+          typename EdgeOp,
+          typename T>
+T transform_reduce_e(raft::handle_t const& handle,
```

Comment:
This function failed to compile for me when T is a simple arithmetic type.  Something in the call stack expects a tuple.  Probably a subtle SFINAE bug on a function in the call stack.

The following failed to compile:
```
    weight_t total_edge_weight = experimental::transform_reduce_e(
      handle_,
      graph_,
      thrust::make_constant_iterator(0),
      thrust::make_constant_iterator(0),
      [] __device__(auto, auto, weight_t wt, auto, auto) { return wt; },
      weight_t{0});
```

Reply:
Thanks, yes this should work for simple arithmetic types or a tuple of simple arithmetic types. Sounds like a bug, I will take a look.

Reply:
This should be fixed by the new commit, and let me know if this still fails.